tests/Test/Xrefcheck/RedirectRequestsSpec.hs

```diff
@@ -0,0 +1,107 @@
+import Data.Map qualified as M
+import Network.HTTP.Types (Status, mkStatus, permanentRedirect308, temporaryRedirect307)
+import Network.HTTP.Types.Header (hLocation)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (Assertion, testCase)
+import Web.Firefly (ToResponse (toResponse), route, run)
+
+import Test.Xrefcheck.UtilRequests
+import Xrefcheck.Progress
+import Xrefcheck.Verify
+
+test_redirectRequests :: TestTree
+test_redirectRequests = testGroup "Redirect response tests"
+  [ testCase "307 no location pass by default" $
+      redirectAssertion
+        temporaryRedirect307
+        "http://127.0.0.1:5000/redirect"
+        Nothing
+        (VerifyResult []),
+    testCase "307 with location pass by default" $
+      redirectAssertion
+        temporaryRedirect307
+        "http://127.0.0.1:5000/redirect"
+        (Just "http://127.0.0.1:5000/other")
+        (VerifyResult []),
+    testCase "302 no location pass by default" $
+      redirectAssertion
+        (mkStatus 302 "Temporary redirect")
+        "http://127.0.0.1:5000/redirect"
+        Nothing
+        (VerifyResult []),
+    testCase "302 with location pass by default" $
+      redirectAssertion
+        (mkStatus 302 "Temporary redirect")
+        "http://127.0.0.1:5000/redirect"
+        (Just "http://127.0.0.1:5000/other")
+        (VerifyResult []),
+    testCase "303 no location pass by default" $
+      redirectAssertion
+        (mkStatus 303 "Temporary redirect")
+        "http://127.0.0.1:5000/redirect"
+        Nothing
+        (VerifyResult []),
+    testCase "303 with location pass by default" $
+      redirectAssertion
+        (mkStatus 303 "Temporary redirect")
+        "http://127.0.0.1:5000/redirect"
+        (Just "http://127.0.0.1:5000/other")
+        (VerifyResult []),
+    testCase "308 no location fail by default" $
+      redirectAssertion
+        permanentRedirect308
+        "http://127.0.0.1:5000/redirect"
+        Nothing
+        (VerifyResult [PermanentRedirectError "http://127.0.0.1:5000/redirect" Nothing]),
+    testCase "308 with location fail by default" $
+      redirectAssertion
+        permanentRedirect308
+        "http://127.0.0.1:5000/redirect"
+        (Just "http://127.0.0.1:5000/other")
+        (VerifyResult [PermanentRedirectError "http://127.0.0.1:5000/redirect" (Just "http://127.0.0.1:5000/other")]),
+    testCase "301 no location fail by default" $
+      redirectAssertion
+        (mkStatus 301 "Permanent redirect")
+        "http://127.0.0.1:5000/redirect"
+        Nothing
+        (VerifyResult [PermanentRedirectError "http://127.0.0.1:5000/redirect" Nothing]),
+    testCase "301 with location fail by default" $
+      redirectAssertion
+        (mkStatus 301 "Permanent redirect")
+        "http://127.0.0.1:5000/redirect"
+        (Just "http://127.0.0.1:5000/other")
+        (VerifyResult [PermanentRedirectError "http://127.0.0.1:5000/redirect" (Just "http://127.0.0.1:5000/other")])
+  ]
```

Comment:
This is only my opinion, but I believe there is no reason to provide that high coverage, since e.g. handling of `302` and `303` in the code is exactly the same. E.g. having 302 w/ loc, 302 w/o loc, 307 w/ loc would be sufficient to cover the temporary redirects.

Using formal terms, we don't usually practice automated "black box" testing when the implementation is only one and concrete.

What do you think?

Reply:
I think that it's OK to check all the possible variants. But it seems that way these tests are written is a bit verbose. We could for example make a separate function which tests two options (location present/absent) for a given code, and pass to this function as a parameter whether these tests should succeed or not. I believe that code will get much shorter this way.

Reply:
If test cases can be turned into one- or two-liners, that approach sounds good.

Reply:
Looks nice :+1: